### ISSUE TITLE
Weights converted but classifying horribly (bug somewhere)

### ISSUE THREAD

**Issue by kentsommer** [kentsommer/keras-inception-resnetV2#1]
Somewhere in either the model or the weight conversion code (from tfslim to Keras with TF dim ordering and Tensorflow backend) there is a bug. I've fixed a couple that I've found already, but currently I'm not able to find out why its not working. I suspect somehow some of the batch norm weights are getting assigned out of order, although, after going over the code multiple times I cannot find any bug that would cause that to happen. 

Also, it should be noted that the tfslim implementation of inception-resnet-v2 is different than the model described in their paper. There is zero explanation for this anywhere that I can find, however, since the checkpoint file is available for the tfslim version, I have copied that model definition for this implementation.

**Comment by kentsommer** [comment]
I have tested the tfslim implementation and it classifies correctly, so the bug is somewhere in my conversion or model definition.

```
import tensorflow as tf
slim = tf.contrib.slim
from PIL import Image
from inception_resnet_v2 import *
import numpy as np

def preprocess_input(x):
    x = np.divide(x, 255.0)
    x = np.subtract(x, 1.0)
    x = np.multiply(x, 2.0)
    return x

checkpoint_file = '../../checkpoints/inception_resnet_v2_2016_08_30.ckpt'
#Load the model
sess = tf.Session()
input_tensor = tf.placeholder(tf.float32, (None, 299, 299, 3))
arg_scope = inception_resnet_v2_arg_scope()
with slim.arg_scope(arg_scope):
  logits, end_points = inception_resnet_v2(input_tensor, is_training=False)
saver = tf.train.Saver()
saver.restore(sess, checkpoint_file)

image = '../../elephant.jpg'
im = Image.open(image).resize((299,299))
im = np.array(im)
im = preprocess_input(im)
im = im.reshape(-1,299,299,3)
predict_values, logit_values = sess.run([end_points['Predictions'], logits], feed_dict={input_tensor: im})
print (np.max(predict_values), np.max(logit_values))
print (np.argmax(predict_values), np.argmax(logit_values))
```

**Comment by kentsommer** [comment]
Could be a relu activation is happening (or not happening) somewhere and causing the issues. I'll try rewriting the model to be explicit about where they happen tomorrow and see if that helps.

**Comment by kentsommer** [comment]
This will have to wait for now as I have more pressing work. I should have time around the 22nd to take a look at this again.

**Comment by jessejohns** [comment]
Kent, just a note, your stem layer looks like Inception-resnet-v1 (Figure 14) rather than Inception-resnet-v2 (Figure 3), which is identical to Inception-v4.  I suspect you've probably transposed what you're working from. :-)

**Comment by kentsommer** [comment]
@jessejohns 

I noticed that as well comparing to their paper, however, this model matches the tfslim implementation found here: https://github.com/tensorflow/models/blob/master/slim/nets/inception_resnet_v2.py

There has not been given any explanation for why their implementation differs from the paper apart from mentioning in their blog post ["we are also releasing a pre-trained instance of the **new** Inception-ResNet-v2"](https://research.googleblog.com/2016/08/improving-inception-and-image.html).

**Comment by jessejohns** [comment]
I see - good point... I think "new" just refers to that it's newer than v3, but still, it's very different than in the paper.  The difference is curious, which could explain why my implementation of their paper does not train very well - Inception v4 does quite well, but the Inception-Resnet-v2 does quite poorly.

**Comment by a-maci** [comment]
Hi Kent, any success with this?
I have also tried to do eval on this model but not getting the stated accuracy.

**Comment by kentsommer** [comment]
@a-maci 

Haven't had any time to spend on this unfortunately. I'm guessing I wont' have much time in the next month or so either as I have a few other pressing projects to finish up. I will keep progress updates here as I find time to look into this.

One odd thing to note... I have noticed accuracy changes depending on how the weights are loaded (using tfslim's restore vs default tf restore) which doesn't make much sense at all. It definitely IS classifying properly in the tfslim implementation and I've triple checked that my model definition matches the tfslim definition exactly, so my current working theory is something wacky with the weight loading / conversion is happening.

**Comment by pengpaiSH** [comment]
@kentsommer Kentsommer, would you provide a weights.h5 file as ```keras.application.InceptionV3``` for Inception-ResNet-V2?

**Comment by kentsommer** [comment]
@pengpaiSH 

That is what I'd like to do, however, as this issue discusses, currently there is a bug (assuming) somewhere so the weights are useless.

**Comment by pengpaiSH** [comment]
@kentsommer Thanks for your feedback! So, only ```inception_v4.h5``` is available for the moment?

**Comment by kentsommer** [comment]
@pengpaiSH 

That is correct.

**Comment by dineshbvadhia** [comment]
@kentsommer dm @titu1994 for fix

**Comment by kentsommer** [comment]
@dineshbvadhia 

@titu1994 and I actually worked together (sort of) to get that weight conversion script he wrote. So the method I used and what he has should be the same, I'll look into if he made some major changes though. 

Thanks for the tip anyways though, if you want an already working version take a look at this: https://github.com/myutwo150/keras-inception-resnet-v2

I won't have much time to work on this repo for some time so for the moment I would suggest using @myutwo150's, he did a great job :)

**Comment by titu1994** [comment]
Sorry I've not been able to spend some more time on porting these weights @kentsommer

On first glance, it seems like @myutwo150 is using the same general technique as what I did for when I was porting the MobileNet weights, but he is using a far better API from Tensorflow to do it (I hacked my way to pure numpy arrays and then manually loaded the weights to the specific layers). Fantastic job, @myutwo150!

**Comment by yuyang-huang** [comment]
Thanks a lot! @kentsommer @titu1994 
I actually learned how to use `NewCheckpointReader` from your code. Your repos were really, really helpful.

**Comment by kentsommer** [comment]
I'm closing this out due to inactivity, and the fact that there is now a working version of resnetV2 ported (not this one) so this repo is not really needed.
